src/lib/common/include/sol-mainloop.h

```diff
@@ -64,11 +64,269 @@ extern "C" {
  * @li @ref Update
  * @li @ref Utils
  * @li @ref WorkerThread
+ * @li @ref patterns
  *
  * Please see the @ref authors page for contact details.
  */
 
 /**
+ * @page patterns Design Patterns
+ * @tableofcontents
+ *
+ * This page will explain the common design patterns employed by Soletta.
+ *
+ * @section streams IO Streams
+ *
+ * Soletta has its own pattern for dealing with input/output streams. This
+ * section will describe how the soletta pattern is implemented and how to use it.
+ *
+ * The aim of this section is to guide the developer the to provide a unified API for all kind of streams, making
+ * it easier for a third party developer to migrate its application to any stream related IO, if necessary.
+ *
+ * The pattern is divided in three categories: The stream configuration, the write API and the read API.
+ *
+ * @subsection config Configuration
+ *
+ * The stream should configured using a struct during its creation, this struct should provide the following fields.
+ *
+ * <ul>
+ * <li> @c tx_size: The write buffer size - if @c 0 it means that is no limit (Data for this buffer is not necessary allocated).
+ * <li> @c rx_size: The read buffer size - If @c 0 a default value will be set (platform specific).
```

Comment:
say that usually this buffer is allocated and the user will be called back with at **most** that amount (upper limit, may be less if less data was available)